contributors/design-proposals/node/accelerator-monitoring.md

```diff
@@ -0,0 +1,101 @@
+
+When cAdvisor starts up, it would read the vendor files in `/sys/bus/pci/devices/*` to see if any NVIDIA devices (vendor ID: `0x10de`) are attached to the node.
+- If no NVIDIA devices are found, this code path would become dormant for the rest of cAdvisor/kubelet lifetime.
+- If NVIDIA devices are found, we would start a goroutine that would check for the presence of NVML by trying to dynamically load it at regular intervals (say every minute or every 5 minutes). We need to do this regular checking instead of doing it just once because it may happen that cAdvisor is started before the nvidia drivers and nvml are installed. Once the NVML dynamic loading succeeds, we would use NVML’s query methods to find out how many devices exist on the node and create a map from their minor numbers to their handles and cache that map. The goroutine would exit at this point.
+
+If we detected the presence of NVML in the previous step, whenever a new container is detected by cAdvisor, cAdvisor would read the `devices.list` file from the container [devices cgroup](https://www.kernel.org/doc/Documentation/cgroup-v1/devices.txt). The `devices.list` file lists the major:minor number of all the devices that the container is allowed to access. If we find any device with major number `195` ([which is the major number assigned to NVIDIA devices](https://github.com/torvalds/linux/blob/v4.13/Documentation/admin-guide/devices.txt#L2583)), we would cache the list of corresponding minor numbers for that container.
+
+During every housekeeping operation, in addition to collecting all the existing metrics, we will use the cached nvidia device minor numbers and the map from minor numbers to device handles to get metrics for GPU devices attached to the container.
+
+The following new metrics would be exposed per container from cAdvisor:
+
+```
+type ContainerStats struct {
+...
+        // Metrics for Accelerators.
+        // Each Accelerator corresponds to one element in the array.
+        Accelerators []AcceleratorStats `json:"accelerators,omitempty"`
+...
+}
+
+type AcceleratorStats struct {
+        // Make of the accelerator (nvidia, amd, google etc.)
+        Make string `json:"make"`
+
+        // Model of the accelerator (tesla-p100, tesla-k80)
+        Model string `json:"model"`
+
+        // ID of the accelerator. device minor number? Or UUID?
+        ID string `json:"id"`
+
+        // Total acclerator memory.
+        // unit: bytes
+        MemoryTotal uint64 `json:"memory_total"`
```

Comment:
nit: either fix the json tags (camel case) or remove them.

Reply:
This section is for what's added to cAdvisor which uses snake_case.

> The following new metrics would be exposed per container from cAdvisor:

I mention below that we will add the same thing to the summary API.

> We will update kubelet’s summary API to also add these metrics.